Gui development

Lehdari opened the pull request.

Lehdari committed:
Streamline loss plot axis autofit code

Lehdari committed:
App: Add adaptive delay to stabilize framerate

Lehdari committed:
App: Implement plot saving

Lehdari committed:
Add time series for all sublosses

Lehdari committed:
Update GraphicsUtils submodule

Some updates required for displaying images in the GUI

Lehdari committed:
Trainer: Add internal DoomGame frame storage

Required for streaming the frame into a OpenGL texture and showing it in the GUI

Lehdari committed:
App: Add window for DoomGame frame visualization

Lehdari committed:
App: Make frame window closeable

Lehdari committed:
Image: Set alpha to 1 when converting from 3 to 4 channels

Lehdari committed:
Model: Add map for image outputs

Lehdari committed:
AutoEncoderModel: Output input1 image

Lehdari committed:
App: Add window for training input1 visualization

Lehdari committed:
Add GuiImageRelay class

GuiImageRelay can be used to render Images from SingleBuffer wrappers into the ImGui

Lehdari committed:
SingleBuffer: Add get() to handles

Lehdari committed:
Image: Add support for grayscale

Lehdari committed:
GuiImageRelay: Add grayscale support

Lehdari committed:
App: Add dynamic drop menu to select model training images

Lehdari committed:
AutoEncoderModel: Add multiple image outputs

Lehdari committed:
AutoEncoderModel: Output all training images via the images interface

Lehdari committed:
GuiImageRelay: Use BGRA for grayscale images too

Using GRAY will cause the images to be rendered using the red channel only

Lehdari committed:
AutoEncoderModel: Remove all OpenCV-related stuff

No more window hell!

Lehdari committed:
App: Use std::map instead of std::unordered_map in ImageRelayMap

Alphabetical ordering in the dropdown menu is kinda nice

Lehdari committed:
App: Plot window improvements

Lehdari committed:
App: Make entire window dockable

Lehdari committed:
App: Make the window resizeable

Lehdari committed:
Refactor GUI code into its own subfolder

Lehdari committed:
Gui: Split windows into polymorphic subclasses

Lehdari committed:
Gui: Add functionality for creating multiple windows

Lehdari committed:
Gui: Move window-specific state to window classes

Lehdari committed:
Gui: Implement layout and window configuration serialization

Hyrtsi commented on include/gui/WindowTypeUtils.hpp:
```diff
@@ -0,0 +1,83 @@
+        return windowTypeName<WINDOW>();
+    GUI_WINDOW_TYPES
+    #undef GUI_WINDOW_TYPE
+}
```
```
/home/eljashyyrynen/source/DooT2/include/gui/WindowTypeUtils.hpp:55:1: warning: control reaches end of non-void function [-Wreturn-type]
   55 | }
      | ^
```

Lehdari replied:
Ended up largely rewriting the window type utils, fixed in the most recent commit.

Lehdari committed:
Gui: Refactor window type utils

- Add window menu labels to the macro
- Replace the runtime typeIds with compile-time ones
- Add windowForEachTypeCallback
